File: vox/typescript/src/rapace/msg-desc-hot.ts
/**
 * Hot-path message descriptor (64 bytes, one cache line).
 *
 * This is the primary descriptor used for frame dispatch.
 * Fits in a single cache line for performance.
 *
 * Binary layout (all little-endian):
 * ```
 * offset  size  field
 * 0       8     msgId: u64
 * 8       4     channelId: u32
 * 12      4     methodId: u32
 * 16      4     payloadSlot: u32
 * 20      4     payloadGeneration: u32
 * 24      4     payloadOffset: u32
 * 28      4     payloadLen: u32
 * 32      4     flags: u32 (FrameFlags bitfield)
 * 36      4     creditGrant: u32
 * 40      8     deadlineNs: u64
 * 48      16    inlinePayload: [u8; 16] (fixed 16 bytes)
 * ```
 */

import { FrameFlags, FrameFlagsType, hasFlag } from "./frame-flags.js";

/** Size of inline payload in bytes. */
export const INLINE_PAYLOAD_SIZE = 16;

/** Sentinel value indicating payload is inline (not in a slot). */
export const INLINE_PAYLOAD_SLOT = 0xffffffff;

/** Sentinel value indicating no deadline. */
export const NO_DEADLINE = 0xffffffffffffffffn;

/** Size of the serialized descriptor in bytes. */
export const MSG_DESC_HOT_SIZE = 64;

export class MsgDescHotError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "MsgDescHotError";
  }
}

/**
 * Hot-path message descriptor.
 */
export class MsgDescHot {
  // Identity (16 bytes)
  /** Unique message ID per session, monotonic. */
  msgId: bigint = 0n;
  /** Logical stream (0 = control channel). */
  channelId: number = 0;
  /** For RPC dispatch, or control verb. */
  methodId: number = 0;

  // Payload location (16 bytes)
  /** Slot index (0xFFFFFFFF = inline). */
  payloadSlot: number = INLINE_PAYLOAD_SLOT;
  /** Generation counter for ABA safety. */
  payloadGeneration: number = 0;
  /** Offset within slot. */
  payloadOffset: number = 0;
  /** Actual payload length. */
  payloadLen: number = 0;

  // Flow control & timing (16 bytes)
  /** Frame flags (EOS, CANCEL, ERROR, etc.). */
  flags: FrameFlagsType = 0;
  /** Credits being granted to peer. */
  creditGrant: number = 0;
  /** Deadline in nanoseconds (monotonic clock). NO_DEADLINE = no deadline. */
  deadlineNs: bigint = NO_DEADLINE;

  // Inline payload for small messages (16 bytes)
  /** When payloadSlot == 0xFFFFFFFF, payload lives here. */
  inlinePayload: Uint8Array = new Uint8Array(INLINE_PAYLOAD_SIZE);

  /**
   * Parse from 64 bytes of raw data.
   */
  static parse(data: Uint8Array): MsgDescHot {
    if (data.length !== MSG_DESC_HOT_SIZE) {
      throw new MsgDescHotError(
        `Invalid size: expected ${MSG_DESC_HOT_SIZE}, got ${data.length}`
      );
    }

    const view = new DataView(data.buffer, data.byteOffset, data.byteLength);
    const desc = new MsgDescHot();

    // Identity (16 bytes)
    desc.msgId = view.getBigUint64(0, true);
    desc.channelId = view.getUint32(8, true);
    desc.methodId = view.getUint32(12, true);

    // Payload location (16 bytes)
    desc.payloadSlot = view.getUint32(16, true);
    desc.payloadGeneration = view.getUint32(20, true);
    desc.payloadOffset = view.getUint32(24, true);
    desc.payloadLen = view.getUint32(28, true);

    // Flow control & timing (16 bytes)
    desc.flags = view.getUint32(32, true);
    desc.creditGrant = view.getUint32(36, true);
    desc.deadlineNs = view.getBigUint64(40, true);

    // Inline payload (16 bytes)
    desc.inlinePayload = data.slice(48, 64);

    return desc;
  }

  /**
   * Serialize to 64 bytes of raw data.
   */
  serialize(): Uint8Array {
    const data = new Uint8Array(MSG_DESC_HOT_SIZE);
    const view = new DataView(data.buffer);

    // Identity (16 bytes)
    view.setBigUint64(0, this.msgId, true);
    view.setUint32(8, this.channelId, true);
    view.setUint32(12, this.methodId, true);

    // Payload location (16 bytes)
    view.setUint32(16, this.payloadSlot, true);
    view.setUint32(20, this.payloadGeneration, true);
    view.setUint32(24, this.payloadOffset, true);
    view.setUint32(28, this.payloadLen, true);

    // Flow control & timing (16 bytes)
    view.setUint32(32, this.flags, true);
    view.setUint32(36, this.creditGrant, true);
    view.setBigUint64(40, this.deadlineNs, true);

    // Inline payload (16 bytes)
    data.set(this.inlinePayload.subarray(0, INLINE_PAYLOAD_SIZE), 48);

    return data;
  }

  /**
   * Returns true if this frame has a deadline set.
   */
  get hasDeadline(): boolean {
    return this.deadlineNs !== NO_DEADLINE;
  }

  /**
   * Returns true if payload is inline (not in a slot).
   */
  get isInline(): boolean {
    return this.payloadSlot === INLINE_PAYLOAD_SLOT;
  }

  /**
   * Returns true if this is a control frame (channel 0).
   */
  get isControl(): boolean {
    return this.channelId === 0;
  }

  /**
   * Returns true if this is a data frame.
   */
  get isData(): boolean {
    return hasFlag(this.flags, FrameFlags.DATA);
  }

  /**
   * Returns true if this is an error response.
   */
  get isError(): boolean {
    return hasFlag(this.flags, FrameFlags.ERROR);
  }

  /**
   * Returns true if this is end of stream.
   */
  get isEos(): boolean {
    return hasFlag(this.flags, FrameFlags.EOS);
  }

  /**
   * Get inline payload data (only valid if isInline).
   */
  getInlinePayloadData(): Uint8Array {
    return this.inlinePayload.subarray(0, this.payloadLen);
  }

  /**
   * Set inline payload from data.
   *
   * @param data - Up to 16 bytes of payload data.
   */
  setInlinePayload(data: Uint8Array): void {
    if (data.length > INLINE_PAYLOAD_SIZE) {
      throw new MsgDescHotError(
        `Inline payload must be at most ${INLINE_PAYLOAD_SIZE} bytes, got ${data.length}`
      );
    }

    // Zero-initialize
    this.inlinePayload.fill(0);

    // Copy data
    this.inlinePayload.set(data);

    this.payloadLen = data.length;
    this.payloadSlot = INLINE_PAYLOAD_SLOT;
  }

  /**
   * Create a debug string representation.
   */
  toString(): string {
    const slotStr =
      this.payloadSlot === INLINE_PAYLOAD_SLOT
        ? "INLINE"
        : String(this.payloadSlot);
    const deadlineStr =
      this.deadlineNs === NO_DEADLINE ? "NONE" : String(this.deadlineNs);

    return `MsgDescHot {
  msgId: ${this.msgId}
  channelId: ${this.channelId}
  methodId: 0x${this.methodId.toString(16).padStart(8, "0")}
  payloadSlot: ${slotStr}
  payloadGeneration: ${this.payloadGeneration}
  payloadOffset: ${this.payloadOffset}
  payloadLen: ${this.payloadLen}
  flags: 0x${this.flags.toString(16)}
  creditGrant: ${this.creditGrant}
  deadlineNs: ${deadlineStr}
  isInline: ${this.isInline}
}`;
  }
}

File: vox/swift-legacy/Sources/Rapace/MsgDescHot.swift
import Foundation

/// Size of inline payload in bytes.
public let inlinePayloadSize: Int = 16

/// Sentinel value indicating payload is inline (not in a slot).
///
/// Spec: `[impl frame.sentinel.values]`
public let inlinePayloadSlot: UInt32 = UInt32.max

/// Sentinel value indicating no deadline.
public let noDeadline: UInt64 = UInt64.max

/// Errors that can occur during MsgDescHot parsing.
public enum MsgDescHotError: Error {
    /// The input data is not exactly 64 bytes.
    case invalidSize(expected: Int, actual: Int)
}

/// Flags carried in each frame descriptor.
public struct FrameFlags: OptionSet, Sendable {
    public let rawValue: UInt32

    public init(rawValue: UInt32) {
        self.rawValue = rawValue
    }

    /// Regular data frame.
    public static let data = FrameFlags(rawValue: 0b0000_0001)
    /// Control frame (channel 0).
    public static let control = FrameFlags(rawValue: 0b0000_0010)
    /// End of stream (half-close).
    public static let eos = FrameFlags(rawValue: 0b0000_0100)
    /// Cancel this channel.
    public static let cancel = FrameFlags(rawValue: 0b0000_1000)
    /// Error response.
    public static let error = FrameFlags(rawValue: 0b0001_0000)
    /// Priority scheduling hint.
    public static let highPriority = FrameFlags(rawValue: 0b0010_0000)
    /// Contains credit grant.
    public static let credits = FrameFlags(rawValue: 0b0100_0000)
    /// Headers/trailers only, no body.
    public static let metadataOnly = FrameFlags(rawValue: 0b1000_0000)
    /// Don't send a reply frame for this request.
    public static let noReply = FrameFlags(rawValue: 0b0001_0000_0000)
}

/// Hot-path message descriptor (64 bytes, one cache line).
///
/// This is the primary descriptor used for frame dispatch.
/// Fits in a single cache line for performance.
///
/// Binary layout (all little-endian):
/// ```
/// offset  size  field
/// 0       8     msgId: UInt64
/// 8       4     channelId: UInt32
/// 12      4     methodId: UInt32
/// 16      4     payloadSlot: UInt32
/// 20      4     payloadGeneration: UInt32
/// 24      4     payloadOffset: UInt32
/// 28      4     payloadLen: UInt32
/// 32      4     flags: UInt32 (FrameFlags bitfield)
/// 36      4     creditGrant: UInt32
/// 40      8     deadlineNs: UInt64
/// 48      16    inlinePayload: [UInt8] (fixed 16 bytes)
/// ```
///
/// Spec: `[impl frame.desc.encoding]`
public struct MsgDescHot: Sendable {
    /// Size of the serialized descriptor in bytes.
    ///
    /// Spec: `[impl frame.desc.size]`
    public static let size: Int = 64

    // Identity (16 bytes)
    /// Unique message ID per session, monotonic.
    public var msgId: UInt64
    /// Logical stream (0 = control channel).
    public var channelId: UInt32
    /// For RPC dispatch, or control verb.
    public var methodId: UInt32

    // Payload location (16 bytes)
    /// Slot index (UInt32.max = inline).
    public var payloadSlot: UInt32
    /// Generation counter for ABA safety.
    public var payloadGeneration: UInt32
    /// Offset within slot.
    public var payloadOffset: UInt32
    /// Actual payload length.
    public var payloadLen: UInt32

    // Flow control & timing (16 bytes)
    /// Frame flags (EOS, CANCEL, ERROR, etc.).
    public var flags: FrameFlags
    /// Credits being granted to peer.
    public var creditGrant: UInt32
    /// Deadline in nanoseconds (monotonic clock). noDeadline = no deadline.
    public var deadlineNs: UInt64

    // Inline payload for small messages (16 bytes)
    /// When payloadSlot == UInt32.max, payload lives here.
    public var inlinePayload: (
        UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8,
        UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8
    )

    /// Create a new descriptor with default values.
    public init() {
        self.msgId = 0
        self.channelId = 0
        self.methodId = 0
        self.payloadSlot = inlinePayloadSlot
        self.payloadGeneration = 0
        self.payloadOffset = 0
        self.payloadLen = 0
        self.flags = []
        self.creditGrant = 0
        self.deadlineNs = noDeadline
        self.inlinePayload = (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)
    }

    /// Parse from 64 bytes of raw data.
    ///
    /// - Parameter data: Exactly 64 bytes of little-endian data.
    /// - Throws: `MsgDescHotError.invalidSize` if data is not 64 bytes.
    public init(from data: Data) throws {
        guard data.count == Self.size else {
            throw MsgDescHotError.invalidSize(expected: Self.size, actual: data.count)
        }

        // Parse using withUnsafeBytes, extracting values to local variables first
        let parsed: (
            msgId: UInt64, channelId: UInt32, methodId: UInt32,
            payloadSlot: UInt32, payloadGeneration: UInt32, payloadOffset: UInt32, payloadLen: UInt32,
            flags: UInt32, creditGrant: UInt32, deadlineNs: UInt64,
            inlinePayload: (UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8,
                           UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8, UInt8)
        ) = data.withUnsafeBytes { buffer in
            let ptr = buffer.baseAddress!

            return (
                msgId: ptr.load(fromByteOffset: 0, as: UInt64.self).littleEndian,
                channelId: ptr.load(fromByteOffset: 8, as: UInt32.self).littleEndian,
                methodId: ptr.load(fromByteOffset: 12, as: UInt32.self).littleEndian,
                payloadSlot: ptr.load(fromByteOffset: 16, as: UInt32.self).littleEndian,
                payloadGeneration: ptr.load(fromByteOffset: 20, as: UInt32.self).littleEndian,
                payloadOffset: ptr.load(fromByteOffset: 24, as: UInt32.self).littleEndian,
                payloadLen: ptr.load(fromByteOffset: 28, as: UInt32.self).littleEndian,
                flags: ptr.load(fromByteOffset: 32, as: UInt32.self).littleEndian,
                creditGrant: ptr.load(fromByteOffset: 36, as: UInt32.self).littleEndian,
                deadlineNs: ptr.load(fromByteOffset: 40, as: UInt64.self).littleEndian,
                inlinePayload: (
                    buffer[48], buffer[49], buffer[50], buffer[51],
                    buffer[52], buffer[53], buffer[54], buffer[55],
                    buffer[56], buffer[57], buffer[58], buffer[59],
                    buffer[60], buffer[61], buffer[62], buffer[63]
                )
            )
        }

        // Assign parsed values to self
        self.msgId = parsed.msgId
        self.channelId = parsed.channelId
        self.methodId = parsed.methodId
        self.payloadSlot = parsed.payloadSlot
        self.payloadGeneration = parsed.payloadGeneration
        self.payloadOffset = parsed.payloadOffset
        self.payloadLen = parsed.payloadLen
        self.flags = FrameFlags(rawValue: parsed.flags)
        self.creditGrant = parsed.creditGrant
        self.deadlineNs = parsed.deadlineNs
        self.inlinePayload = parsed.inlinePayload
    }

    /// Serialize to 64 bytes of raw data.
    ///
    /// - Returns: 64 bytes of little-endian data.
    public func serialize() -> Data {
        var data = Data(count: Self.size)

        data.withUnsafeMutableBytes { buffer in
            let ptr = buffer.baseAddress!

            // Identity (16 bytes)
            ptr.storeBytes(of: msgId.littleEndian, toByteOffset: 0, as: UInt64.self)
            ptr.storeBytes(of: channelId.littleEndian, toByteOffset: 8, as: UInt32.self)
            ptr.storeBytes(of: methodId.littleEndian, toByteOffset: 12, as: UInt32.self)

            // Payload location (16 bytes)
            ptr.storeBytes(of: payloadSlot.littleEndian, toByteOffset: 16, as: UInt32.self)
            ptr.storeBytes(of: payloadGeneration.littleEndian, toByteOffset: 20, as: UInt32.self)
            ptr.storeBytes(of: payloadOffset.littleEndian, toByteOffset: 24, as: UInt32.self)
            ptr.storeBytes(of: payloadLen.littleEndian, toByteOffset: 28, as: UInt32.self)

            // Flow control & timing (16 bytes)
            ptr.storeBytes(of: flags.rawValue.littleEndian, toByteOffset: 32, as: UInt32.self)
            ptr.storeBytes(of: creditGrant.littleEndian, toByteOffset: 36, as: UInt32.self)
            ptr.storeBytes(of: deadlineNs.littleEndian, toByteOffset: 40, as: UInt64.self)

            // Inline payload (16 bytes)
            buffer[48] = inlinePayload.0
            buffer[49] = inlinePayload.1
            buffer[50] = inlinePayload.2
            buffer[51] = inlinePayload.3
            buffer[52] = inlinePayload.4
            buffer[53] = inlinePayload.5
            buffer[54] = inlinePayload.6
            buffer[55] = inlinePayload.7
            buffer[56] = inlinePayload.8
            buffer[57] = inlinePayload.9
            buffer[58] = inlinePayload.10
            buffer[59] = inlinePayload.11
            buffer[60] = inlinePayload.12
            buffer[61] = inlinePayload.13
            buffer[62] = inlinePayload.14
            buffer[63] = inlinePayload.15
        }

        return data
    }

    /// Returns true if this frame has a deadline set.
    public var hasDeadline: Bool {
        deadlineNs != noDeadline
    }

    /// Returns true if payload is inline (not in a slot).
    public var isInline: Bool {
        payloadSlot == inlinePayloadSlot
    }

    /// Returns true if this is a control frame (channel 0).
    public var isControl: Bool {
        channelId == 0
    }

    /// Get inline payload as Data (only valid if isInline).
    public var inlinePayloadData: Data {
        let bytes: [UInt8] = [
            inlinePayload.0, inlinePayload.1, inlinePayload.2, inlinePayload.3,
            inlinePayload.4, inlinePayload.5, inlinePayload.6, inlinePayload.7,
            inlinePayload.8, inlinePayload.9, inlinePayload.10, inlinePayload.11,
            inlinePayload.12, inlinePayload.13, inlinePayload.14, inlinePayload.15
        ]
        return Data(bytes.prefix(Int(payloadLen)))
    }

    /// Set inline payload from Data.
    ///
    /// - Parameter data: Up to 16 bytes of payload data.
    /// - Precondition: data.count <= 16
    public mutating func setInlinePayload(_ data: Data) {
        precondition(data.count <= inlinePayloadSize, "Inline payload must be at most \(inlinePayloadSize) bytes")

        // Zero-initialize
        inlinePayload = (0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0)

        // Copy data
        if data.count > 0 { inlinePayload.0 = data[data.startIndex] }
        if data.count > 1 { inlinePayload.1 = data[data.startIndex + 1] }
        if data.count > 2 { inlinePayload.2 = data[data.startIndex + 2] }
        if data.count > 3 { inlinePayload.3 = data[data.startIndex + 3] }
        if data.count > 4 { inlinePayload.4 = data[data.startIndex + 4] }
        if data.count > 5 { inlinePayload.5 = data[data.startIndex + 5] }
        if data.count > 6 { inlinePayload.6 = data[data.startIndex + 6] }
        if data.count > 7 { inlinePayload.7 = data[data.startIndex + 7] }
        if data.count > 8 { inlinePayload.8 = data[data.startIndex + 8] }
        if data.count > 9 { inlinePayload.9 = data[data.startIndex + 9] }
        if data.count > 10 { inlinePayload.10 = data[data.startIndex + 10] }
        if data.count > 11 { inlinePayload.11 = data[data.startIndex + 11] }
        if data.count > 12 { inlinePayload.12 = data[data.startIndex + 12] }
        if data.count > 13 { inlinePayload.13 = data[data.startIndex + 13] }
        if data.count > 14 { inlinePayload.14 = data[data.startIndex + 14] }
        if data.count > 15 { inlinePayload.15 = data[data.startIndex + 15] }

        payloadLen = UInt32(data.count)
        payloadSlot = inlinePayloadSlot
    }
}

extension MsgDescHot: CustomDebugStringConvertible {
    public var debugDescription: String {
        """
        MsgDescHot {
            msgId: \(msgId)
            channelId: \(channelId)
            methodId: \(methodId)
            payloadSlot: \(payloadSlot == inlinePayloadSlot ? "INLINE" : String(payloadSlot))
            payloadGeneration: \(payloadGeneration)
            payloadOffset: \(payloadOffset)
            payloadLen: \(payloadLen)
            flags: \(flags.rawValue)
            creditGrant: \(creditGrant)
            deadlineNs: \(deadlineNs == noDeadline ? "NONE" : String(deadlineNs))
            isInline: \(isInline)
        }
        """
    }
}
